File: app/assets/javascripts/factories/claim_form.js
angular.module("moBilling.factories.claimForm", [])

    .factory("ClaimForm", function () {
        var ClaimForm = function (attributes) {
            angular.extend(this, attributes);
        };

        ClaimForm.fromResource = function (resource) {
            var attributes = {};

            attributes.id = resource.id;
            attributes.photo_id = resource.photo_id;
            attributes.patient_name = resource.patient_name;
            attributes.hospital = resource.hospital;
            attributes.referring_physician = resource.referring_physician;
            attributes.diagnosis = resource.diagnosis;
            attributes.most_responsible_physician = resource.most_responsible_physician;
            attributes.admission_on = resource.admission_on;
            attributes.first_seen_admission = (resource.admission_on === resource.first_seen_on);
            attributes.first_seen_on = resource.first_seen_on;
            attributes.last_seen_on = resource.last_seen_on;
            attributes.last_seen_discharge = resource.last_seen_discharge;

            return new this(attributes);
        };

        ClaimForm.prototype.toResource = function () {
            var attributes = {};

            attributes.id = this.id;
            attributes.photo_id = this.photo_id;
            attributes.patient_name = this.patient_name;
            attributes.hospital = this.hospital;
            attributes.referring_physician = this.referring_physician;
            attributes.diagnosis = this.diagnosis;
            attributes.most_responsible_physician = this.most_responsible_physician;
            attributes.admission_on = this.admission_on;
            attributes.first_seen_on = this.first_seen_admission ? this.admission_on : this.first_seen_on;
            attributes.last_seen_on = this.last_seen_on;
            attributes.last_seen_discharge = this.last_seen_discharge;

            return attributes;
        };

        return ClaimForm;
    });

File: app/forms/claim_form.rb
class ClaimForm
  include ActiveModel::Model
  include Virtus.model
  include ValidationScopes

  def self.model_params
    return [
            [:photo_id, String],
            [:status, String],
            [:user_id, String]
           ]
  end

  def self.details_params
    return [
            [:specialty, String],
            [:patient_name, String],
            [:hospital, String],
            [:referring_physician, String],
            [:most_responsible_physician, :bool],
            [:admission_on, String],
            [:first_seen_on, String],
            [:first_seen_consult, :bool],
            [:last_seen_on, String],
            [:last_seen_discharge, :bool],
            [:icu_transfer, :bool],
            [:consult_type, String],
            [:consult_time_in, String],
            [:consult_time_out, String],
            [:consult_premium_visit, String],
            [:consult_premium_first, :bool],
            [:consult_premium_travel, :bool],
            [:patient_number, String],
            [:patient_province, String],
            [:patient_birthday, String],
            [:patient_sex, String],
            [:referring_laboratory, String],
            [:payment_program, String],
            [:payee, String],
            [:manual_review_indicator, :bool],
            [:service_location, String],
            [:last_code_generation, String],
           ]
  end

  def self.helper_params
    return [
            [:comment, String],           # new comment
            [:num_comments, Integer],     # number of comments when editing started
            [:consult_time_type, String], # if set, can return premium counts
            [:unread_comments, Integer],  # number of unread comments for this user.   Set to 0 to set read flags
           ]
  end

  def self.scalar_params
    return model_params + details_params + helper_params
  end

  def self.array_params
    return [
            [:diagnoses, Array, [[:name, String]]],
            [:daily_details, Array, DailyDetailForm.all_params],
           ]
  end

  def self.all_params
    return scalar_params + array_params
  end

  def self.param_names(list)
    return list.map do |name, type, array|
      if type == Array
        { name => param_names(array) }
      else
        name
      end
    end
  end

  def self.all_param_names
    param_names(all_params)
  end

  all_params.each do |name, type|
    if type == :bool
      attribute name, Axiom::Types::Boolean
    else
      attribute name, type
    end
  end
  attribute :user, User

end
